### ISSUE TITLE
G28 — Multi-Domain Adapter: All-4 Combination

### ISSUE THREAD

**Issue by samscarrow** [git-scarrow/fast-associative-memory#41]
## Gauntlet 28 — Multi-Domain Adapter: All-4 Combination

**Phase:** Phase 4 – Exploration
**Kill Condition:** Combined adapter trained on all 4 fine-grained domains underperforms the best individual single-domain adapter on average across all evaluation datasets (in-domain + cross-domain).

**Dependencies:** G22–G25 (single-domain adapter results), G26 (cross-domain baselines), G27 (top-2 multi-domain reference).

---

### Objective

Create a script `benchmarks/g28_multi_adapter_all4.py` that:

1. **Trains a multi-domain MetricAdapter on the combined training data of ALL 4 fine-grained domains:**
   - CUB-200 Birds (G22)
   - Stanford Cars (G23)
   - FGVC Aircraft (G24)
   - Oxford Flowers-102 (G25)
   - Architecture: `MetricAdapter(input_dim=1024, output_dim=256, hidden_dim=512)`
   - Same OHNM triplet loss protocol as `train_adapter_triplet.py`
   - 5 epochs, batch_size=128, lr=1e-4, margin=0.2
   - Training data: concatenate all 4 DINOv2 feature caches, remap labels so they don't collide
   - Save weights to `adapter_multi_all4.pt`

2. **Evaluates the all-4 multi-domain adapter on ALL evaluation datasets:**
   - **In-domain** (all 4 domains): top-1 accuracy, prototype count, condensation ratio at v=0.80
   - **Cross-domain** (CIFAR-100 + ImageNet-R): accuracy and drop vs no-adapter baseline (G19/G26 protocol)

3. **Computes the kill condition:**
   - Average accuracy across all 6 evaluation datasets (4 in-domain + 2 cross-domain) for the all-4 multi-domain adapter
   - Average accuracy across all 6 evaluation datasets for the best individual single-domain adapter
   - If multi-domain average < best single-domain average → kill condition met
   - Also compare against G27 top-2 multi-domain adapter (informational, not kill)

### CLI Interface

```
python benchmarks/g28_multi_adapter_all4.py \
  --cache-birds ./data/cub200 \
  --cache-cars ./data/stanford_cars \
  --cache-aircraft ./data/fgvc_aircraft \
  --cache-flowers ./data/flowers102 \
  --cifar-cache ./feature_cache_vitb14 \
  --imagenet-r-cache ./feature_cache_imagenet_r_vitl14 \
  --adapter-birds adapter_cub200_birds.pt \
  --adapter-cars adapter_stanford_cars.pt \
  --adapter-aircraft adapter_fgvc_aircraft.pt \
  --adapter-flowers adapter_flowers102.pt \
  --adapter-top2 adapter_multi_top2.pt \
  [--epochs 5] [--batch-size 128] [--lr 1e-4] [--margin 0.2] [--seed 42] [--device auto]
```

- `--cache-{domain}`: path to feature cache directory for each domain (must contain `*train*.pt` and `*test*.pt`)
- `--adapter-{domain}`: paths to single-domain adapter weights for comparison
- `--adapter-top2`: optional path to G27 multi-top2 adapter for informational comparison

### Expected Output

```
G28 — Multi-Domain Adapter: All-4 Combination
=========================================================================
  Dataset            Best Single   Multi-Top2   Multi-All4   Δ vs Best (pp)
  -------------------------------------------------------------------------
  Birds (in)         XX.XX%        XX.XX%       XX.XX%       +X.XX
  Cars (in)          XX.XX%        XX.XX%       XX.XX%       +X.XX
  Aircraft (in)      XX.XX%        XX.XX%       XX.XX%       +X.XX
  Flowers (in)       XX.XX%        XX.XX%       XX.XX%       +X.XX
  CIFAR-100 (cross)  XX.XX%        XX.XX%       XX.XX%       +X.XX
  ImageNet-R (cross) XX.XX%        XX.XX%       XX.XX%       +X.XX
  -------------------------------------------------------------------------
  AVERAGE            XX.XX%        XX.XX%       XX.XX%       +X.XX
=========================================================================

Kill condition: Multi-all4 avg XX.XX% vs best single avg XX.XX%
→ PASS / KILL
```

### Implementation Details

- Import from existing modules: `adapter.MetricAdapter`, `fast_associative_memory.FastAssociativeMemory`, `nstp.NSTPController`
- Reuse `load_features()` from `benchmarks/gauntlet_19_cross_domain.py` for loading feature caches
- Reuse training loop from `train_adapter_triplet.py` (OHNM miner + triplet loss)
- When concatenating training data from all 4 domains, remap labels so they don't collide (e.g., domain A: 0..N₁-1, domain B: N₁..N₁+N₂-1, etc.)
- If any adapter file is missing (single-domain or top-2), print a warning and skip that comparison column
- If G27 `benchmarks/g27_multi_adapter_top2.py` has been merged, feel free to share helpers. If not, implement standalone.
- Exit code 0 if pass, 1 if kill

### Files to Create
- `benchmarks/g28_multi_adapter_all4.py`

### Files to Reference (do not modify)
- `adapter.py` — `MetricAdapter` class
- `train_adapter_triplet.py` — OHNM training protocol reference
- `fast_associative_memory.py` — `FastAssociativeMemory` class
- `nstp.py` — `NSTPController` class
- `benchmarks/gauntlet_19_cross_domain.py` — feature loading + evaluation protocol
